ONSdigital/rm-collection-exercise-service
# Fix event validation when adding and updating events

# Motivation and Context

Currently the validation is not flexible enough to allow validation
on both update and creation. This PR makes the validation work with
both create events and update events. Also making sure all but two
 business rules (24hrs apart and dates are in the future) are captured 
within the validation.

# What has changed



Add 'complete' event validation when both updating and creating events

# How to test?



Ran unit tests, ITs, ATs

Can use this response operation branch if you prefer clicking around over api calls>> https://github.com/ONSdigital/response-operations-ui/pull/223

Double check business rules covered:
* The user can only input Main Print Selection date (MPS) that ~~is at least 24 hours~~ before the Go Live Date 
* The user can only input a Go Live Date that is after Main Print Selection but ~~at least 24 hours~~ before the Return By Date
* The user can only enter a Return By Date that is after the Go Live Date but ~~at least 24 hours~~ before the Collection Exercise End Date
* The user can only enter a Collection Exercise End Date that is after the Return By Date
* ~~All Mandatory dates must also be in the future~~

* ~~All Optional event dates can not be set in the past apart from Reference Start and End dates & Employment dates~~
* The user can only enter a Reminder Date that is ~~at least 24 hours~~ after the Go Live Date and ~~at least 24 hours~~ before the Collection Exercise End Date
* Second reminder date must be ~~at least 24 hours~~after the first reminder date and ~~at least 24 hours~~ before Collection end date
* Third reminder date must be ~~at least 24 hours~~ after the second reminder date and ~~at least 24 hours~~ before Collection end date
* The reference period end date must always be after the reference period start date
* Employment date has no validation

* The mandatory business event dates are editable but they cannot be changed once the state of a collection exercise is Ready for Live (and all other states that indicate the CE is/has been executed)

# Links
